executor/hash_table.go

```diff
@@ -120,6 +144,66 @@ func (c *hashRowContainer) GetMatchedRows(probeKey uint64, probeRow chunk.Row, h
 	return matchedRows, err
 }
 
+func (c *hashRowContainer) GetAllMatchedRows(probeHCtx *hashContext, probeSideRow chunk.Row,
+	probeKeyNullBits *bitmap.ConcurrentBitmap, matched []chunk.Row, needCheckBuildRowPos, needCheckProbeRowPos []int) ([]chunk.Row, error) {
+	// for NAAJ probe row with null, we should match them with all build rows.
+	var (
+		ok        bool
+		err       error
+		innerPtrs []chunk.RowPtr
+	)
+	c.hashTable.Iter(
+		func(_ uint64, e *entry) {
+			entryAddr := e
+			for entryAddr != nil {
+				innerPtrs = append(innerPtrs, entryAddr.ptr)
+				entryAddr = entryAddr.next
+			}
+		})
+	matched = matched[:0]
+	if len(innerPtrs) == 0 {
+		return matched, nil
+	}
+	var mayMatchedRow chunk.Row
+	for _, ptr := range innerPtrs {
+		mayMatchedRow, c.chkBuf, err = c.rowContainer.GetRowAndAppendToChunk(ptr, c.chkBuf)
+		if err != nil {
+			return nil, err
+		}
+		// all built bucket rows come from hash table, their bitmap are all nil (doesn't contain any null). so
+		// we could only use the probe null bits to filter valid rows.
+		if probeKeyNullBits != nil && len(probeHCtx.naKeyColIdx) > 1 {
+			// if len(probeHCtx.naKeyColIdx)=1
+			//     that means the NA-Join probe key is directly a (null) <-> (fetch all buckets), nothing to do.
+			// else like
+			//	   (null, 1, 2), we should use the not-null probe bit to filter rows. Only fetch rows like
+			//     (  ? , 1, 2), that exactly with value as 1 and 2 in the second and third join key column.
+			needCheckProbeRowPos = needCheckProbeRowPos[:0]
+			needCheckBuildRowPos = needCheckBuildRowPos[:0]
+			keyColLen := len(c.hCtx.naKeyColIdx)
+			for i := 0; i < keyColLen; i++ {
+				// since all bucket is from hash table (Not Null), so the buildSideNullBits check is eliminated.
+				if probeKeyNullBits.UnsafeIsSet(i) {
+					continue
+				}
+				needCheckBuildRowPos = append(needCheckBuildRowPos, c.hCtx.naKeyColIdx[i])
+				needCheckProbeRowPos = append(needCheckProbeRowPos, probeHCtx.naKeyColIdx[i])
+			}
+			// check the idxs-th value of the join columns.
```

Comment:
What does `idxs` here mean?

Reply:
join column（a,b,c）

find the non-null bit, for example, b,c 
use the b,c's column index to fetch the value from the chunk.
the idxs (means b,c's column index)

Reply:
Looks like `needCheckProbeRowPos` and `needCheckBuildRowPos` can be generated outside the loop since it only related to `probeKeyNullBits`?